### PR TITLE
i40e driver

### PR DESCRIPTION
**Pull request by aroulin**
Adding support for the 40GbE Intel cards.

### CODE REVIEW

File: dp/drivers/i40e.c [hunk 1]
```diff
@@ -81,12 +81,18 @@
...
+	rte_eth_allmulticast_enable(dev->port);
 }
 
 static void dev_infos_get(struct ix_rte_eth_dev *dev, struct ix_rte_eth_dev_info *dev_info)
 {
-	panic("dev_infos_get not implemented\n");
+	struct rte_eth_dev_info dpdk_dev_info;
+
+	rte_eth_dev_info_get(dev->port, &dpdk_dev_info);
+	dev_info->nb_rx_fgs = 128;
```
Comment (aroulin) [inline dp/drivers/i40e.c]: Not sure about that line. Copied from `ixgbe.c`. Is it related to RSS or is it IX-specific?

File: dp/drivers/i40e.c [hunk 2]
```diff
@@ -34,6 +34,12 @@
 #include <rte_config.h>
 #include <rte_ethdev.h>
 
+/* DPDK i40e includes */
+#define I40E_MAC_X722_VF 8 //FIXME
```
Comment (aroulin) [inline dp/drivers/i40e.c]: w/o that line:

``````
In file included from drivers/i40e.c:40:0:
../deps/dpdk/drivers/net/i40e/i40e_ethdev.h: In function 'i40e_get_vsi_from_adapter':
../deps/dpdk/drivers/net/i40e/i40e_ethdev.h:617:53: error: 'I40E_MAC_X722_VF' undeclared (first use in this function)
  if (hw->mac.type == I40E_MAC_VF || hw->mac.type == I40E_MAC_X722_VF) ```
``````

One way to fix it is c6a14a4e744d

File: dp/drivers/i40e.c [hunk 3]
```diff
@@ -128,10 +158,186 @@ static int reta_update(struct ix_rte_eth_dev *dev, struct rte_eth_rss_reta *reta
...
+	 * we require the number of descriptors to be a power of 2.
+	 */
+	if (nb_desc & (nb_desc - 1))
+		return -EINVAL;
+
+	/* NOTE: This is a hack, but it's the only way to support late/lazy
+	 * queue setup in DPDK; a feature that IX depends on. */
+	rte_eth_devices[dev->port].data->nb_rx_queues = queue_idx + 1;
+
+	ret = rte_eth_rx_queue_setup(dev->port, queue_idx, nb_desc, numa_node, &rx_conf, dpdk_pool);
```
Comment (aroulin) [inline dp/drivers/i40e.c]: Using global `rx_conf` here. Need to check it works as I accidentally was using a zero-filled default conf (and it worked surprisingly!). We can also use `NULL` and have dpdk use a default one.

File: dp/drivers/i40e.c [hunk 4]
```diff
@@ -128,10 +158,186 @@ static int reta_update(struct ix_rte_eth_dev *dev, struct rte_eth_rss_reta *reta
...
+	log_err("queue_setup : phys %p\n", rxq->ring_physaddr);
+	/* TODO: this is now done in dev_start (need access to dpdk's queue) */
+	//rxq->reg_idx = drxq->reg_idx;
+	//rxq->rdt_reg_addr = (uint32_t*) drxq->qrx_tail;
+
+	rxq->erxq.poll = i40e_rx_poll;
+	dev->data->rx_queues[queue_idx] = &rxq->erxq;
+	rxq->erxq.drv_rx_queue = rxq;
+	/* release the dpdk memory location and all its buffers*/
+	//i40e_dev_rx_queue_release(drxq);
```
Comment (aroulin) [inline dp/drivers/i40e.c:531]: Do we want to free dpdk's queue?

File: dp/drivers/i40e.c (hunk 1 above)
Comment (prekageo) [reply dp/drivers/i40e.c]: It is related to RSS. It is hardware-specific. ixgbe has 128 flow groups. How many does the i40e has? Can we get this number using a DPDK API call?

File: dp/drivers/i40e.c (hunk 2 above)
Comment (prekageo) [reply dp/drivers/i40e.c]: Maybe you are not `#include`ing something? Copy-pasting a `#define` should be our last resort. Maybe you need to say `#define X722_SUPPORT` (move it from the gcc command line here)? But is it necessary?

File: dp/drivers/i40e.c (hunk 3 above)
Comment (prekageo) [reply dp/drivers/i40e.c]: Using `NULL` sounds fine. Unless, IX needs to pass some specific configuration options that are not DPDK's default options. Is that the case?

File: dp/drivers/i40e.c (hunk 4 above)
Comment (prekageo) [reply dp/drivers/i40e.c:531]: That's a good idea. Unless it breaks something in a weird way. If it works correctly, then you could to do the same in the ixgbe driver later on.

File: dp/drivers/i40e.c (hunk 1 above)
Comment (aroulin) [reply dp/drivers/i40e.c]: Do I understand correctly that this is the RETA size? We can get it with `dev_infos` http://dpdk.org/doc/api/structrte__eth__dev__info.html#a8f02d328521addc69743ec44f124065c

Comment (prekageo) [reply dp/drivers/i40e.c]: Yes, that's correct. I guess we should get this information from it. I also hope that DPDK returns 128 for the ixgbe.

File: dp/drivers/i40e.c [hunk 5]
```diff
@@ -158,7 +158,7 @@ static void dev_infos_get(struct ix_rte_eth_dev *dev, struct ix_rte_eth_dev_info
 	struct rte_eth_dev_info dpdk_dev_info;
 
 	rte_eth_dev_info_get(dev->port, &dpdk_dev_info);
-	dev_info->nb_rx_fgs = dpdk_dev_info.reta_size;
```
Comment (aroulin) [inline dp/drivers/i40e.c]: This returns 512.
It seems it causes the following function to overflow and zero out fields of the queue (drv_rx_queue is null after it).
https://github.com/aroulin/ix/blob/i40e-driver-release/dp/core/ethdev.c#L260

Need to debug that. With 128 it works fine.

Comment (prekageo) [reply dp/drivers/i40e.c]: Check out this: https://github.com/ix-project/ix/blob/bd4dd0ba200413716071561a3c3f2994159c665d/inc/ix/ethfg.h#L38

There is an upper limit on the number of flow groups that IX supports. Try to raise it,

**Comment by aroulin** [comment]
`ETH_RSS_RETA_NUM_ENTRIES` would need to be dynamically changed depending on the driver (`ixgbe` or `i40e`).
https://github.com/ix-project/ix/blob/bd4dd0ba200413716071561a3c3f2994159c665d/inc/ix/ethdev.h#L225

File: dp/drivers/i40e.c [hunk 6]
```diff
@@ -354,9 +354,30 @@ static void promiscuous_disable(struct ix_rte_eth_dev *dev)
...
+	r_reta_conf[0].mask = reta_conf->mask_lo;
+	r_reta_conf[1].mask = reta_conf->mask_hi;
+
+	for (i = 0; i < RTE_RETA_GROUP_SIZE; ++i) {
+		r_reta_conf[0].reta[i] = reta_conf->reta[i];
+		r_reta_conf[1].reta[i] = reta_conf->reta[RTE_RETA_GROUP_SIZE + i];
+	}
+
+	spin_lock(&i40e_dev_lock);
+	ret = r_dev.dev_ops->reta_update(&r_dev, r_reta_conf, ETH_RSS_RETA_NUM_ENTRIES);
```
Comment (aroulin) [inline dp/drivers/i40e.c]: `ETH_RSS_RETA_NUM_ENTRIES` would need to be dynamically changed depending on the driver (`ixgbe` or `i40e`).
https://github.com/ix-project/ix/blob/bd4dd0ba200413716071561a3c3f2994159c665d/inc/ix/ethdev.h#L225

File: dp/drivers/i40e.c [hunk 7]
```diff
@@ -354,9 +354,30 @@ static void promiscuous_disable(struct ix_rte_eth_dev *dev)
 	rte_eth_promiscuous_disable(dev->port);
 }
 
+static DEFINE_SPINLOCK(i40e_dev_lock);
+
 static int reta_update(struct ix_rte_eth_dev *dev, struct rte_eth_rss_reta *reta_conf)
 {
-	panic("reta update not implemented\n");
+	int ret, i;
+	struct rte_eth_rss_reta_entry64 r_reta_conf[2];
```
Comment (aroulin) [inline dp/drivers/i40e.c]: I think this `2` is a magic number that should be changed to `ETH_RSS_RETA_NUM_ENTRIES / 64`

File: inc/ix/ethqueue.h [hunk 8]
```diff
@@ -117,6 +118,8 @@ struct eth_tx_queue {
 
 	int (*reclaim)(struct eth_tx_queue *tx);
 	int (*xmit)(struct eth_tx_queue *tx, int nr, struct mbuf **mbufs);
+
+	void *drv_tx_queue; /*pointer to driver queue data*/
```
Comment (prekageo) [inline inc/ix/ethqueue.h]: I understand that you use the field `drv_tx_queue` to access the `tx_queue` from the `eth_tx_queue`. You don't need this field. Take a look at the definition of `tx_queue`:

```
struct tx_queue {
    struct eth_tx_queue etxq;
    ...
}
```

`tx_queue` contains an `eth_tx_queue`. That means that given an `eth_tx_queue` you can get a `tx_queue` by using a `container_of` trick/macro.

File: inc/ix/ethqueue.h [hunk 9]
```diff
@@ -59,6 +59,7 @@ struct eth_rx_queue {
 	DEFINE_BITMAP(assigned_fgs, ETH_MAX_NUM_FG);
 
 	struct ix_rte_eth_dev *dev;
+	void *drv_rx_queue; /*pointer to driver queue data*/
```
Comment (prekageo) [inline inc/ix/ethqueue.h]: Same comment as below.

File: inc/ix/ethqueue.h (hunk 8 above)
Comment (aroulin) [reply inc/ix/ethqueue.h]: right, it's fixed now.

File: inc/ix/ethqueue.h (hunk 9 above)
Comment (aroulin) [reply inc/ix/ethqueue.h]: fixed too.

File: dp/drivers/i40e.c (hunk 6 above)
Comment (aroulin) [reply dp/drivers/i40e.c]: In order to handle a RETA table of size 512, this struct would need to be modified too:
https://github.com/ix-project/ix/blob/master/inc/ix/ethdev.h#L296.

File: inc/ix/ethdev.h [hunk 10]
```diff
@@ -222,7 +222,7 @@ struct ix_rte_eth_rss_conf {
 #define ETH_RSS_IPV6_UDP    0x0080 /**< IPv6/UDP packet. */
 #define ETH_RSS_IPV6_UDP_EX 0x0100 /**< IPv6/UDP with extension headers. */
 /* Definitions used for redirection table entry size */
-#define ETH_RSS_RETA_NUM_ENTRIES 128
+#define ETH_RSS_RETA_NUM_ENTRIES 512
 #define ETH_RSS_RETA_MAX_QUEUE   16
```
Comment (aroulin) [inline inc/ix/ethdev.h]: Is `ETH_RSS_RETA_MAX_QUEUE` IX-specific or related to the driver? It is used here:
https://github.com/aroulin/ix/blob/i40e-driver-release/dp/core/ethdev.c#L100

i40e DPDK driver says `dev_info.max_rx_queues` is 320 which is quite higher than 16.

Comment (prekageo) [reply inc/ix/ethdev.h]: Changing the `#define` from 128 to 512 doesn't seem like the right thing to do. Remember, that IX must work both with ixgbe (which has 128 entries) and i40e (which has 512 entries). You should remove the `#define` and implement a different configuration scheme.

Comment (prekageo) [reply inc/ix/ethdev.h]: As for the `ETH_RSS_RETA_MAX_QUEUE`, IX wants an RX queue per thread. I am not sure that we have a good reason for capping the number of queues to 16 except that our servers have 16 HTs. Don't worry about it for this pull request. We might want to make it more flexible (support servers with more HTs) later on, though.

Comment (aroulin) [reply inc/ix/ethdev.h]: of course for the `#define ETH_RSS_RETA_NUM_ENTRIES`. I was just trying it to see if it works. It does give `0 <= fg_id < 512`

File: inc/ix/ethdev.h [hunk 11]
```diff
@@ -291,12 +295,9 @@ struct rte_eth_vmdq_mirror_conf {
  * A structure used to configure Redirection Table of  the Receive Side
  * Scaling (RSS) feature of an Ethernet port.
  */
-struct rte_eth_rss_reta {
-	/** First 64 mask bits indicate which entry(s) need to updated/queried. */
-	uint64_t mask_lo;
-	/** Second 64 mask bits indicate which entry(s) need to updated/queried. */
-	uint64_t mask_hi;
-	uint8_t reta[ETH_RSS_RETA_NUM_ENTRIES];  /**< 128 RETA entries*/
+struct ix_rte_eth_rss_reta_entry64 {
```
Comment (aroulin) [inline inc/ix/ethdev.h]: This is the same definition as in DPDK
http://dpdk.org/doc/api/structrte__eth__rss__reta__entry64.html

File: dp/core/ethfg.c [hunk 12]
```diff
@@ -150,12 +150,13 @@ static int eth_fg_assign_single_to_cpu(int fg_id, int cpu, struct rte_eth_rss_re
 		ret = 1;
 	}
 
-	if (fg->idx >= 64)
-		rss_reta->mask_hi |= ((uint64_t) 1) << (fg->idx - 64);
-	else
-		rss_reta->mask_lo |= ((uint64_t) 1) << fg->idx;
+	if (*rss_reta == NULL) {
+		*rss_reta = malloc(ETH_RSS_RETA_NUM_GROUPS(fg->eth) * sizeof(struct ix_rte_eth_rss_reta_entry64));
```
Comment (aroulin) [inline dp/core/ethfg.c]: Need to figure out exactly where we want to allocate the dynamically-sized `rss_reta` array + check return value.

File: dp/core/dpdk.c [hunk 13]
```diff
@@ -59,7 +59,6 @@ int dpdk_init(void)
 	const int pool_size = 32768;
 
 	optind = 0;
-	internal_config.no_hugetlbfs = 1;
```
Comment (prekageo) [inline dp/core/dpdk.c:62]: Apparently the initialization of the i40e in DPDK requires huge pages, so I understand the reason for this change. However, I've noticed that the boot time of IX has increased after removing this line. Can you investigate if we can initialize the i40e without removing this line?

File: dp/drivers/i40e.c [hunk 14]
```diff
@@ -0,0 +1,1003 @@
...
+	if (ret < 0)
+		return ret;
+
+	ix_reta_conf->rss_key = reta_conf.rss_key;
+	ix_reta_conf->rss_hf = reta_conf.rss_hf;
+
+	return ret;
+}
+
+static void mac_addr_add(struct ix_rte_eth_dev *dev, struct eth_addr *mac_addr, uint32_t index, uint32_t vmdq)
```
Comment (prekageo) [inline dp/drivers/i40e.c]: There is code duplication. This function is the same as https://github.com/ix-project/ix/blob/master/dp/drivers/ixgbe.c#L857. Can you please identify the duplicated functions between these two files? Then, please move them to https://github.com/ix-project/ix/blob/master/dp/drivers/common.c. I'd like to see a separate pull request where you refactor ixgbe.c in that way.

**Review by prekageo — CHANGES_REQUESTED** [review]

File: dp/core/dpdk.c (hunk 13 above)
Comment (aroulin) [reply dp/core/dpdk.c:62]: I'll investigate.

**Review by prekageo — APPROVED** [review]

**Pull request merged**